File: vantage-cli/src/tracker.ts
import { randomUUID } from "node:crypto";
import { VERSION } from "./_version.js";
import { bus } from "./event-bus.js";
import { calculateCost } from "./pricing.js";
import { countTokens } from "./optimizer.js";
import { getAgent } from "./agents/registry.js";

export interface TrackerConfig {
  apiKey: string;
  apiBase: string;
  batchSize: number;
  flushInterval: number;
  privacy: "full" | "strict" | "anonymized" | "local-only";
  debug: boolean;
}

interface DashboardEvent {
  event_id: string;
  provider: string;
  model: string;
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
  total_cost_usd: number;
  latency_ms: number;
  environment: string;
  agent_name: string;
  team: string;
}

export class Tracker {
  private queue: DashboardEvent[] = [];
  private timer: ReturnType<typeof setInterval> | null = null;
  private config: TrackerConfig;
  private sentIds = new Set<string>();   // confirmed-delivered IDs
  private queuedIds = new Set<string>(); // IDs currently in the send queue
  private exitRegistered = false;
  private lastSavedTokens = 0;
  private lastPromptText = "";

  constructor(config: TrackerConfig) {
    this.config = config;
    this.setupListeners();
  }

  private resolveProvider(agent: string): string {
    const map: Record<string, string> = {
      claude: "anthropic",
      codex: "openai",
      gemini: "google",
      aider: "anthropic",
      chatgpt: "openai",
    };
    return map[agent] || "other";
  }

  private setupListeners(): void {
    bus.on("prompt:optimized", (data) => {
      this.lastSavedTokens = data.savedTokens;
    });

    bus.on("prompt:submitted", (data) => {
      this.lastPromptText = data.prompt;
    });

    bus.on("agent:completed", (data) => {
      const agent = getAgent(data.agent);
      const model = agent?.defaultModel ?? "unknown";
      const outputTokens = countTokens(data.outputText);
      // Use actual prompt text for input tokens when available
      const inputTokens = this.lastPromptText
        ? countTokens(this.lastPromptText)
        : Math.ceil(outputTokens * 0.25); // Fallback: ~25% of output (more realistic)
      this.lastPromptText = "";
      const costUsd = calculateCost(model, inputTokens, outputTokens);

      // Calculate saved cost from optimization
      const savedTokens = this.lastSavedTokens;
      this.lastSavedTokens = 0; // reset for next prompt
      const savedCost = savedTokens > 0 ? calculateCost(model, savedTokens, 0) : 0;

      bus.emit("cost:calculated", {
        agent: data.agent,
        model,
        inputTokens,
        outputTokens,
        costUsd,
        savedUsd: savedCost,
      });

      const event: DashboardEvent = {
        event_id: randomUUID(),
        provider: this.resolveProvider(data.agent),
        model,
        prompt_tokens: inputTokens,
        completion_tokens: outputTokens,
        total_tokens: inputTokens + outputTokens,
        total_cost_usd: costUsd,
        latency_ms: data.durationMs,
        environment: "production",
        agent_name: data.agent,
        team: "",
      };

      this.enqueue(event);
    });
  }

  enqueue(event: DashboardEvent): void {
    if (this.config.privacy === "local-only") return;

    if (this.config.privacy === "strict" || this.config.privacy === "anonymized") {
      // Strip identifying info — keep only numeric metrics
      event.agent_name = "anonymous";
      event.team = "";
    }

    // Deduplication guard — check both confirmed-sent and in-queue sets
    const eventKey = event.event_id;
    if (this.sentIds.has(eventKey) || this.queuedIds.has(eventKey)) return;
    this.queuedIds.add(eventKey);
    // Prune queuedIds more aggressively to avoid memory spikes in long sessions
    if (this.queuedIds.size > 1000) {
      const arr = Array.from(this.queuedIds);
      this.queuedIds = new Set(arr.slice(-500));
    }

    this.queue.push(event);

    if (this.queue.length >= this.config.batchSize) {
      this.flush().catch(() => {});
    }
  }

  async flush(): Promise<void> {
    if (this.queue.length === 0) return;
    if (!this.config.apiKey) return;

    if (this.config.apiBase && !this.config.apiBase.startsWith("https://")) {
      if (this.config.debug) console.warn("[vantage] Skipping tracking: API base is not HTTPS");
      return;
    }

    const batch = this.queue.splice(0, this.queue.length);
    const base = this.config.apiBase || "https://api.vantageaiops.com";
    const url = `${base}/v1/events/batch`;

    try {
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${this.config.apiKey}`,
        },
        body: JSON.stringify({
          events: batch,
          sdk_version: `vantage-cli-${VERSION}`,
          sdk_language: "typescript",
        }),
        signal: AbortSignal.timeout(15000),
      });

      bus.emit("cost:reported", { success: response.ok });

      if (response.ok) {
        // Only mark as confirmed-delivered after a 2xx response
        for (const e of batch) {
          this.sentIds.add(e.event_id);
          this.queuedIds.delete(e.event_id);
        }
        // Cap sentIds to prevent unbounded growth in long sessions
        if (this.sentIds.size > 1000) {
          const arr = Array.from(this.sentIds);
          this.sentIds = new Set(arr.slice(-500));
        }
      } else {
        // Never log response body — could contain echoed credentials
        console.error(`  [vantage] Dashboard sync failed: HTTP ${response.status}`);
        // Re-queue events so they're not lost on server-side errors (BUG 14 fix:
        // sentIds no longer contains these IDs so the filter works correctly)
        const unsent = batch.filter((e) => !this.sentIds.has(e.event_id));
        this.queue.unshift(...unsent);
      }
    } catch (err) {
      bus.emit("cost:reported", { success: false });
      if (this.config.debug) {
        console.error("[vantage] Failed to send events:", err);
      }
      // Re-queue failed events (skip any already confirmed sent)
      const unsent = batch.filter((e) => !this.sentIds.has(e.event_id));
      this.queue.unshift(...unsent);
    }
  }

  start(): void {
    if (this.timer) return;
    this.timer = setInterval(() => {
      this.flush().catch(() => {});
    }, this.config.flushInterval);

    // Flush on beforeExit only — SIGTERM/SIGINT are handled by the REPL's
    // shutdown() which already calls tracker.flush() before process.exit().
    // Registering our own exit handlers here would race with the REPL's cleanup,
    // causing double-exit and skipping the session summary print.
    if (!this.exitRegistered) {
      this.exitRegistered = true;
      process.on("beforeExit", () => this.flush().catch(() => {}));
    }
  }

  stop(): void {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
  }
}

File: vantage-agent/vantage_agent/tracker.py
"""
tracker.py — Dashboard telemetry client.

Batches cost/usage events and sends them to the VantageAI backend API.
Respects privacy modes: full, strict, anonymized, local-only.
Ported from vantage-cli/src/tracker.ts.
"""
from __future__ import annotations

import threading
import time
import uuid
from dataclasses import dataclass, field
from typing import Any

import httpx

from .cost_tracker import SessionCost

__version__ = "0.1.0"


@dataclass
class TrackerConfig:
    api_key: str = ""
    api_base: str = "https://api.vantageaiops.com"
    batch_size: int = 10
    flush_interval: float = 30.0  # seconds
    privacy: str = "full"  # full | strict | anonymized | local-only
    debug: bool = False


@dataclass
class DashboardEvent:
    event_id: str
    provider: str
    model: str
    prompt_tokens: int
    completion_tokens: int
    total_tokens: int
    total_cost_usd: float
    latency_ms: int
    environment: str = "cli"
    agent_name: str = "vantage-agent"
    team: str = "default"
    session_id: str = ""


PROVIDER_MAP = {
    "claude": "anthropic",
    "codex": "openai",
    "gemini": "google",
    "aider": "anthropic",
    "chatgpt": "openai",
}


class Tracker:
    """Batched telemetry sender for the VantageAI dashboard."""

    def __init__(self, config: TrackerConfig) -> None:
        self.config = config
        self._queue: list[DashboardEvent] = []
        self._lock = threading.Lock()
        self._timer: threading.Timer | None = None
        self._running = False

    def start(self) -> None:
        if self.config.privacy == "local-only" or not self.config.api_key:
            return
        self._running = True
        self._schedule_flush()

    def stop(self) -> None:
        self._running = False
        if self._timer:
            self._timer.cancel()
        self.flush()

    def record(
        self,
        model: str,
        input_tokens: int,
        output_tokens: int,
        cost_usd: float,
        latency_ms: int,
        agent_name: str = "vantage-agent",
    ) -> None:
        """Queue a usage event."""
        event = DashboardEvent(
            event_id=str(uuid.uuid4()),
            provider=PROVIDER_MAP.get(agent_name, "anthropic"),
            model=model,
            prompt_tokens=input_tokens,
            completion_tokens=output_tokens,
            total_tokens=input_tokens + output_tokens,
            total_cost_usd=cost_usd,
            latency_ms=latency_ms,
            agent_name=agent_name,
        )
        with self._lock:
            self._queue.append(event)
            if len(self._queue) >= self.config.batch_size:
                self._do_flush()

    def flush(self) -> None:
        with self._lock:
            self._do_flush()

    def _do_flush(self) -> None:
        if not self._queue or not self.config.api_key:
            return
        batch = self._queue[:]
        self._queue.clear()

        events = []
        for e in batch:
            data: dict[str, Any] = {
                "event_id": e.event_id,
                "provider": e.provider,
                "model": e.model,
                "prompt_tokens": e.prompt_tokens,
                "completion_tokens": e.completion_tokens,
                "total_tokens": e.total_tokens,
                "total_cost_usd": e.total_cost_usd,
                "latency_ms": e.latency_ms,
                "environment": e.environment,
                "agent_name": e.agent_name,
                "team": e.team,
            }
            if self.config.privacy == "strict":
                data.pop("agent_name", None)
            events.append(data)

        try:
            url = f"{self.config.api_base}/v1/events/batch"
            httpx.post(
                url,
                json={"events": events},
                headers={
                    "Authorization": f"Bearer {self.config.api_key}",
                    "Content-Type": "application/json",
                    "User-Agent": f"vantage-agent/{__version__}",
                },
                timeout=10,
            )
            if self.config.debug:
                print(f"  [tracker] flushed {len(events)} events")
        except Exception as exc:
            if self.config.debug:
                print(f"  [tracker] flush error: {exc}")

    def _schedule_flush(self) -> None:
        if not self._running:
            return
        self._timer = threading.Timer(self.config.flush_interval, self._flush_and_reschedule)
        self._timer.daemon = True
        self._timer.start()

    def _flush_and_reschedule(self) -> None:
        self.flush()
        self._schedule_flush()
